test/dataframerow.jl

```diff
@@ -116,4 +122,16 @@ module TestDataFrameRow
                    c=["A", "B", "C", "A", "B", missing])
     @test copy(DataFrameRow(df, 1)) == (a = 1, b = 2.0, c = "A")
     @test isequal(copy(DataFrameRow(df, 2)), (a = 2, b = missing, c = "B"))
+
+    # iteration and collect
+    ref = ["a", "b", "c"]
+    df = DataFrame(permutedims(ref))
+    dfr = df[1, :]
+    @test collect(dfr) == ref
```

Comment:
Also test the type? When you say the eltype is `Any`, I guess this applies only to heterogeneous column types, right? It would be useful to cover that case too.

(BTW, `Vector` doesn't promote AFAICT, as it uses a comprehension.)

Reply:
I will fix it and test it. Actually it should be narrowest type (as in comprehensions).

What is problematic (and confused me) that in `Matrix(df)` we use `T = reduce(promote_type, eltypes(df))` and probably we should use `T = reduce(typejoin, eltypes(df))`. Should I fix it?

Reply:
See discussion at https://github.com/JuliaData/DataFrames.jl/pull/1641. Maybe we should use promotion here too.

Reply:
As discussed 😄 in #1641, if someone wants a promotion to `T` one should use `Vector{T}(dfr)` in my opinion.

Reply:
So should we use `promote_type` here too for consistency with #1641?

Reply:
I thought that this was exactly the point we wanted to have a difference here.

Essentially this is the same as:

```
julia> collect((1, 1.0, true))
3-element Array{Real,1}:
    1
    1.0
 true

julia> [v for v in (1, 1.0, true)]
3-element Array{Real,1}:
    1
    1.0
 true
```

But I am not too attached to what we have now. But it we want to do the promotion here we should define `eltype` (and this is exactly what I have removed some 2 commits ago 😄).

Reply:
Maybe we should keep `collect` as-is for consistency with tuples, but have `convert` use promotion to be consistent with data frames?

Reply:
This is what we do. But it is implemented in https://github.com/JuliaData/DataFrames.jl/pull/1640.